File: swift-api/Sources/NovelAIAPI/Client/Payload.swift
import Foundation

// MARK: - Payload Construction Helpers
//
// Internal functions for building JSON payloads to send to the NovelAI API.
// Uses [String: Any] dictionaries for flexible conditional field inclusion,
// serialized via JSONSerialization.

// MARK: - Base Payload

/// Build the base payload structure for image generation.
///
/// Constructs the top-level payload with `input`, `model`, `action`, and the
/// `parameters` dictionary containing all standard generation settings.
func buildBasePayload(
    _ params: GenerateParams,
    seed: UInt32,
    negativePrompt: String
) -> [String: Any] {
    let extraNoiseSeed = Int(seed == 0 ? MAX_SEED : seed - 1)
    let v5 = params.model.isV5
    var parameters: [String: Any] = [
        // V5: params_version 4 (same as the official site)
        "params_version": v5 ? 4 : 3,
        "width": params.width,
        "height": params.height,
        "scale": params.scale,
        "sampler": params.sampler.rawValue,
        "steps": params.steps,
        "n_samples": 1,
        "ucPreset": 2,
        "qualityToggle": true,
        "autoSmea": false,
        "dynamic_thresholding": false,
        "controlnet_strength": 1,
        "legacy": false,
        "add_original_image": true,
        "cfg_rescale": params.cfgRescale,
        // V5: noise_schedule is always karras on the official site
        "noise_schedule": v5 ? NoiseSchedule.karras.rawValue : params.noiseSchedule.rawValue,
        "legacy_v3_extend": false,
        "skip_cfg_above_sigma": NSNull(), // Explicit JSON null — API expects this key present with null value
        "use_coords": true,
        "legacy_uc": false,
        "normalize_reference_strength_multiple": true,
        "inpaintImg2ImgStrength": 0.85,
        "seed": Int(seed),
        "extra_noise_seed": extraNoiseSeed,
        "negative_prompt": negativePrompt,
        "deliberate_euler_ancestral_bug": false,
        "prefer_brownian": true,
        "stream": "msgpack",
        "image_format": "png",
    ]
    if params.transparentBackground {
        // Transparency itself comes from the prompt tag; these are the site's hints
        parameters["straight_alpha"] = true
        parameters["tag_hint_transparent_background"] = true
    }

    return [
        "input": params.effectivePrompt,
        "model": params.model.rawValue,
        "action": params.action.rawValue,
        "parameters": parameters,
        "use_new_shared_trial": true,
    ]
}

// MARK: - Img2Img Parameters

/// Apply img2img-specific parameters to the payload.
///
/// When action is `img2img` and a source image is provided, adds the base64
/// image, strength, noise, and extra_noise_seed fields.
func applyImg2ImgParams(
    _ payload: inout [String: Any],
    params: GenerateParams
) throws {
    guard params.action == .img2img, let sourceImage = params.sourceImage else {
        return
    }

    var parameters = payload["parameters"] as? [String: Any] ?? [:]

    // Resize source image to match output dimensions to avoid server errors with oversized images
    let imageBase64 = try resizeImageForImg2Img(sourceImage, targetWidth: params.width, targetHeight: params.height)
    parameters["image"] = imageBase64
    parameters["strength"] = params.img2imgStrength
    parameters["noise"] = params.img2imgNoise

    payload["parameters"] = parameters
}

// MARK: - Infill/Inpaint Parameters

/// Apply infill (inpaint) parameters to the payload.
///
/// Handles model name suffixing, mask resizing, cache key generation,
/// and all inpaint-specific parameter fields.
func applyInfillParams(
    _ payload: inout [String: Any],
    params: GenerateParams
) throws {
    guard params.action == .infill,
          let sourceImage = params.sourceImage,
          params.mask != nil else {
        return
    }

    // Switch to the inpainting model (V5 curated uses the 4.5 curated one)
    payload["model"] = params.model.inpaintModel

    var parameters = payload["parameters"] as? [String: Any] ?? [:]

    // Resize source image to target dimensions (same as img2img)
    let sourceImageBase64 = try resizeImageForImg2Img(
        sourceImage, targetWidth: params.width, targetHeight: params.height
    )

    // Resize mask to 1/8 dimensions
    let maskBuffer = try getImageBuffer(params.mask!)
    let resizedMask = try resizeMaskImage(maskBuffer, targetWidth: params.width, targetHeight: params.height)
    let maskBase64 = resizedMask.base64EncodedString()

    // Validate mask_strength is present
    guard let maskStrength = params.maskStrength else {
        throw NovelAIError.validation("mask_strength is required for infill action")
    }

    let hybridStrength = params.hybridImg2imgStrength ?? maskStrength
    let hybridNoise = params.hybridImg2imgNoise ?? 0

    // Set inpaint parameters
    parameters["image"] = sourceImageBase64
    parameters["mask"] = maskBase64
    parameters["strength"] = hybridStrength
    parameters["noise"] = hybridNoise
    parameters["add_original_image"] = false
    parameters["inpaintImg2ImgStrength"] = maskStrength
    parameters["img2img"] = [
        "strength": maskStrength,
        "color_correct": params.inpaintColorCorrect,
    ] as [String: Any]

    payload["parameters"] = parameters
}

// MARK: - Vibe Transfer Parameters

/// Apply vibe transfer parameters to the payload.
///
/// Adds reference image encodings, strength values, and information extracted
/// values when vibes are present.
func applyVibeParams(
    _ payload: inout [String: Any],
    vibeEncodings: [String],
    vibeStrengths: [Double]?,
    vibeInfoList: [Double]
) {
    guard !vibeEncodings.isEmpty else {
        return
    }

    var parameters = payload["parameters"] as? [String: Any] ?? [:]

    parameters["reference_image_multiple"] = vibeEncodings
    parameters["reference_strength_multiple"] = vibeStrengths
    parameters["reference_information_extracted_multiple"] = vibeInfoList
    parameters["normalize_reference_strength_multiple"] = true

    payload["parameters"] = parameters
}

// MARK: - Character Reference Parameters

/// Apply character reference parameters to the payload.
///
/// Adds director reference images, descriptions, strength values, and
/// streaming configuration for character reference generation.
func applyCharRefParams(
    _ payload: inout [String: Any],
    charRefs: ProcessedCharacterReferences
) {
    var parameters = payload["parameters"] as? [String: Any] ?? [:]

    parameters["director_reference_images"] = charRefs.images
    parameters["director_reference_descriptions"] = charRefs.descriptions.map { $0.toDictionary() }
    parameters["director_reference_information_extracted"] = charRefs.infoExtracted
    parameters["director_reference_strength_values"] = charRefs.strengthValues
    parameters["director_reference_secondary_strength_values"] = charRefs.secondaryStrengthValues

    payload["parameters"] = parameters
}

// MARK: - V4 Prompt Structure

/// Build the v4_prompt structure for the payload.
///
/// Constructs the nested caption dictionary with base_caption and char_captions,
/// along with use_coords and use_order flags.
func buildV4PromptStructure(
    prompt: String,
    charCaptions: [[String: Any]]
) -> [String: Any] {
    return [
        "caption": [
            "base_caption": prompt,
            "char_captions": charCaptions,
        ] as [String: Any],
        "use_coords": true,
        "use_order": true,
    ]
}

/// Build the v4_negative_prompt structure for the payload.
///
/// Constructs the nested caption dictionary for negative prompts with
/// base_caption, char_captions, and legacy_uc flag.
func buildV4NegativePromptStructure(
    negativePrompt: String,
    charNegativeCaptions: [[String: Any]]
) -> [String: Any] {
    return [
        "caption": [
            "base_caption": negativePrompt,
            "char_captions": charNegativeCaptions,
        ] as [String: Any],
        "legacy_uc": false,
    ]
}

// MARK: - Character Prompts

/// Apply character-specific prompts (use_coords) to the payload.
///
/// When characters are configured, enables coordinate mode and adds the
/// characterPrompts array with prompt, negative prompt, center position,
/// and enabled flag for each character.
func applyCharacterPrompts(
    _ payload: inout [String: Any],
    params: GenerateParams
) {
    var parameters = payload["parameters"] as? [String: Any] ?? [:]

    let characters = params.characters ?? []

    if !characters.isEmpty {
        parameters["use_coords"] = true
    }
    parameters["characterPrompts"] = characters.map { char -> [String: Any] in
        [
            "prompt": char.prompt,
            "uc": char.negativePrompt,
            "center": ["x": char.centerX, "y": char.centerY] as [String: Double],
            "enabled": true,
        ]
    }

    payload["parameters"] = parameters
}

// MARK: - Full V4 Prompt Application

/// Apply the complete V4 prompt and negative prompt structures to the payload.
///
/// Combines buildV4PromptStructure and buildV4NegativePromptStructure,
/// setting both v4_prompt and v4_negative_prompt in the parameters.
func applyV4PromptStructures(
    _ payload: inout [String: Any],
    prompt: String,
    negativePrompt: String,
    charCaptions: [[String: Any]],
    charNegativeCaptions: [[String: Any]]
) {
    var parameters = payload["parameters"] as? [String: Any] ?? [:]

    parameters["v4_prompt"] = buildV4PromptStructure(
        prompt: prompt,
        charCaptions: charCaptions
    )
    parameters["v4_negative_prompt"] = buildV4NegativePromptStructure(
        negativePrompt: negativePrompt,
        charNegativeCaptions: charNegativeCaptions
    )

    payload["parameters"] = parameters
}

File: rust-api/src/client/payload.rs
use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;

use crate::constants;
use crate::error::Result;
use crate::schemas::*;
use crate::utils;

/// Build the base generation payload as a JSON value.
pub fn build_base_payload(
    params: &GenerateParams,
    seed: u64,
    negative_prompt: &str,
) -> serde_json::Value {
    serde_json::json!({
        "input": params.prompt,
        "model": params.model.as_str(),
        "action": params.action.as_str(),
        "parameters": {
            "params_version": 3,
            "width": params.width,
            "height": params.height,
            "scale": params.scale,
            "sampler": params.sampler.as_str(),
            "steps": params.steps,
            "n_samples": 1,
            "ucPreset": 0,
            "qualityToggle": false,
            "autoSmea": false,
            "dynamic_thresholding": false,
            "controlnet_strength": 1,
            "legacy": false,
            "add_original_image": true,
            "cfg_rescale": params.cfg_rescale,
            "noise_schedule": params.noise_schedule.as_str(),
            "legacy_v3_extend": false,
            "skip_cfg_above_sigma": null,
            "use_coords": false,
            "legacy_uc": false,
            "normalize_reference_strength_multiple": true,
            "inpaintImg2ImgStrength": 1,
            "seed": seed,
            "negative_prompt": negative_prompt,
            "deliberate_euler_ancestral_bug": false,
            "prefer_brownian": true,
        },
        "use_new_shared_trial": true,
    })
}

/// Apply Img2Img parameters to the payload.
pub fn apply_img2img_params(
    payload: &mut serde_json::Value,
    params: &GenerateParams,
    seed: u64,
) -> Result<()> {
    if let GenerateAction::Img2Img { ref source_image, strength, noise } = params.action {
        let b64 = utils::image::resize_image_for_img2img(source_image, params.width, params.height)?;
        let extra_seed = if seed == 0 {
            constants::MAX_SEED as u64
        } else {
            seed - 1
        };
        payload["parameters"]["image"] = serde_json::Value::String(b64);
        payload["parameters"]["strength"] = serde_json::json!(strength);
        payload["parameters"]["noise"] = serde_json::json!(noise);
        payload["parameters"]["extra_noise_seed"] = serde_json::json!(extra_seed);
        payload["parameters"]["stream"] = serde_json::Value::String("msgpack".to_string());
        payload["parameters"]["image_format"] = serde_json::Value::String("png".to_string());
    }
    Ok(())
}

/// Apply Infill/Inpaint parameters to the payload.
pub fn apply_infill_params(
    payload: &mut serde_json::Value,
    params: &GenerateParams,
    seed: u64,
) -> Result<()> {
    if let GenerateAction::Infill {
        ref source_image,
        ref mask,
        mask_strength,
        color_correct,
        hybrid_strength,
        hybrid_noise,
    } = params.action
    {
        // Append -inpainting suffix (prevent duplicates)
        let model_str = params.model.as_str();
        let model_name = if model_str.ends_with("-inpainting") {
            model_str.to_string()
        } else {
            format!("{}-inpainting", model_str)
        };
        payload["model"] = serde_json::Value::String(model_name);

        // Source image: resize to target dimensions (same as img2img)
        let source_base64 = utils::image::resize_image_for_img2img(
            source_image, params.width, params.height
        )?;
        let source_buffer = BASE64.decode(source_base64.as_bytes())
            .map_err(|e| crate::error::NovelAIError::Image(format!("Failed to decode resized source image: {}", e)))?;

        // Mask: resize to 1/8 of target dimensions
        let mask_buffer = utils::image::get_image_buffer(mask)?;
        let resized_mask =
            utils::mask::resize_mask_image(&mask_buffer, params.width, params.height)?;
        let mask_base64 = BASE64.encode(&resized_mask);

        // Cache secret keys (SHA256)
        let image_cache_key = utils::mask::calculate_cache_secret_key(&source_buffer);
        let mask_cache_key = utils::mask::calculate_cache_secret_key(&resized_mask);

        // Strength parameters
        let effective_hybrid_strength = hybrid_strength.unwrap_or(mask_strength);
        let effective_hybrid_noise = hybrid_noise.unwrap_or(0.0);

        let extra_seed = if seed == 0 {
            constants::MAX_SEED as u64
        } else {
            seed - 1
        };

        payload["parameters"]["image"] = serde_json::Value::String(source_base64);
        payload["parameters"]["mask"] = serde_json::Value::String(mask_base64);
        payload["parameters"]["strength"] = serde_json::json!(effective_hybrid_strength);
        payload["parameters"]["noise"] = serde_json::json!(effective_hybrid_noise);
        payload["parameters"]["add_original_image"] = serde_json::json!(false);
        payload["parameters"]["extra_noise_seed"] = serde_json::json!(extra_seed);
        payload["parameters"]["inpaintImg2ImgStrength"] = serde_json::json!(mask_strength);
        payload["parameters"]["img2img"] = serde_json::json!({
            "strength": mask_strength,
            "color_correct": color_correct,
        });
        payload["parameters"]["image_cache_secret_key"] =
            serde_json::Value::String(image_cache_key);
        payload["parameters"]["mask_cache_secret_key"] =
            serde_json::Value::String(mask_cache_key);
        payload["parameters"]["image_format"] = serde_json::Value::String("png".to_string());
        payload["parameters"]["stream"] = serde_json::Value::String("msgpack".to_string());
    }

    Ok(())
}

/// Apply Vibe Transfer parameters to the payload.
pub fn apply_vibe_params(
    payload: &mut serde_json::Value,
    vibe_encodings: &[String],
    vibe_strengths: &Option<Vec<f64>>,
    vibe_info_list: &[f64],
) {
    if vibe_encodings.is_empty() {
        return;
    }
    payload["parameters"]["reference_image_multiple"] = serde_json::json!(vibe_encodings);
    payload["parameters"]["reference_strength_multiple"] = serde_json::json!(vibe_strengths);
    payload["parameters"]["reference_information_extracted_multiple"] =
        serde_json::json!(vibe_info_list);
    payload["parameters"]["normalize_reference_strength_multiple"] = serde_json::json!(true);
}

/// Apply Character Reference parameters to the payload.
pub fn apply_char_ref_params(
    payload: &mut serde_json::Value,
    char_ref_data: &crate::utils::charref::ProcessedCharacterReferences,
) {
    payload["parameters"]["director_reference_images"] =
        serde_json::json!(char_ref_data.images);
    payload["parameters"]["director_reference_descriptions"] =
        serde_json::json!(char_ref_data.descriptions);
    payload["parameters"]["director_reference_information_extracted"] =
        serde_json::json!(char_ref_data.info_extracted);
    payload["parameters"]["director_reference_strength_values"] =
        serde_json::json!(char_ref_data.strength_values);
    payload["parameters"]["director_reference_secondary_strength_values"] =
        serde_json::json!(char_ref_data.secondary_strength_values);
    payload["parameters"]["stream"] = serde_json::Value::String("msgpack".to_string());
    payload["parameters"]["image_format"] = serde_json::Value::String("png".to_string());
}

/// Build V4 prompt structure for the payload.
pub fn build_v4_prompt_structure(
    payload: &mut serde_json::Value,
    prompt: &str,
    negative_prompt: &str,
    char_captions: &[CaptionDict],
    char_negative_captions: &[CaptionDict],
) {
    let char_caps: Vec<serde_json::Value> = char_captions
        .iter()
        .map(|c| {
            serde_json::json!({
                "char_caption": c.char_caption,
                "centers": c.centers.iter().map(|center| {
                    serde_json::json!({"x": center.x, "y": center.y})
                }).collect::<Vec<_>>(),
            })
        })
        .collect();

    let char_neg_caps: Vec<serde_json::Value> = char_negative_captions
        .iter()
        .map(|c| {
            serde_json::json!({
                "char_caption": c.char_caption,
                "centers": c.centers.iter().map(|center| {
                    serde_json::json!({"x": center.x, "y": center.y})
                }).collect::<Vec<_>>(),
            })
        })
        .collect();

    let has_characters = !char_caps.is_empty();
    payload["parameters"]["v4_prompt"] = serde_json::json!({
        "caption": {
            "base_caption": prompt,
            "char_captions": char_caps,
        },
        "use_coords": has_characters,
        "use_order": true,
    });

    payload["parameters"]["v4_negative_prompt"] = serde_json::json!({
        "caption": {
            "base_caption": negative_prompt,
            "char_captions": char_neg_caps,
        },
        "legacy_uc": false,
    });
}

/// Apply character prompts (use_coords) to the payload.
pub fn apply_character_prompts(
    payload: &mut serde_json::Value,
    char_configs: &[CharacterConfig],
) {
    if !char_configs.is_empty() {
        payload["parameters"]["use_coords"] = serde_json::json!(true);
    }
    let prompts: Vec<serde_json::Value> = char_configs
        .iter()
        .map(|c| {
            serde_json::json!({
                "prompt": c.prompt,
                "uc": c.negative_prompt,
                "center": {"x": c.center_x, "y": c.center_y},
                "enabled": true,
            })
        })
        .collect();
    payload["parameters"]["characterPrompts"] = serde_json::json!(prompts);
}
